### PR TITLE
Implement Scope Support

### PR DESCRIPTION
**Pull request by dshanske**
This starts the process of adding scope support to the plugin. What it does is create a class that defines a scope as a collection of WordPress capabilities.

It then defines a Scopes object that stores the scope definitions and adds a filter to map_meta_cap which, if there is an indieauth process going on, it will compare the list of capabilities required to the ones in the scope, and if any of the requires ones are missing, it will revoke permission.

This means that any plugin can just use current_user_can instead of managing scopes, as they currently have to do. 

There is an issue in that Microsub does not have any specific capabilities built into WordPress. Will be looking into that in the future, but until then, full permissions are granted to all users, as the read capability is used by all. This would require some custom capabilities that could be added to roles. 

This is instantiated within the file, but if merged, I will change that in the PR I agreed to do after this and the other pending PR are merged.

### CODE REVIEW

**Comment by dshanske** [comment]
Thinking of changing the meta cap filter to fire later. Thoughts?

File: includes/class-indieauth-scopes.php [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+				)
+			)
+		);
+
+		$this->register_scope(
+			new IndieAuth_Scope(
+				'delete',
+				__( 'Allows the application to delete or undelete posts', 'indieauth' ),
+				array(
+					'delete_posts',
```
Comment (pfefferle) [inline includes/class-indieauth-scopes.php]: doesn't this also require `edit_posts` or is this implicit?

File: includes/class-indieauth-scopes.php [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+				__( 'Allows the application read access to Microsub channels', 'indieauth' ),
+				array(
+					'read',
+				)
+			)
+		);
+
+		$this->register_scope(
+			new IndieAuth_Scope(
+				'follow',
```
Comment (pfefferle) [inline includes/class-indieauth-scopes.php]: Shouldn't this be added by the Microsub plugin? Shouldn't the IndieAuth plugin as generic as possible?

**Review by pfefferle — APPROVED** [review]
I had only a look at the code and din't test the functionality.

(perhaps you could also some more php-doc-blocks)

**Comment by pfefferle** [comment]
Nice idea btw.!

File: includes/class-indieauth-scopes.php (hunk 1 above)
Comment (dshanske) [reply includes/class-indieauth-scopes.php]: edit_posts would be the permission for update

File: includes/class-indieauth-scopes.php (hunk 2 above)
Comment (dshanske) [reply includes/class-indieauth-scopes.php]: It is...these are all the known scopes for IndieAuth right now.

Comment (pfefferle) [reply includes/class-indieauth-scopes.php]: But the comment is very specific with Microsub.

**Comment by dshanske** [comment]
Working on adding a few docblocks, then I will merge and rebase the other pr

**Comment by dshanske** [comment]
Adding some unit tests as well, before merge, just to check for issues.

**Pull request merged**